Automatically turn on dynamo in cond

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108028

A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

ydwu4 committed:
Automatically turn on dynamo in cond

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,16 +30,136 @@
+        # We need to turn off the is_fx_tracing_flag. Remove this flag check from dyanmo
+        # once we are confident fx tracing works with dynamo.
+        torch.fx._symbolic_trace._is_fx_tracing_flag = False
+        # We need to enlarge the dynamo cache size limit because cond's frame is
+        # hitted with different inputs frequently by tests
+        torch._dynamo.config.cache_size_limit = 256
+        yield
```
If this is just for testing, then we should just reset() before/after each of the cond tests. Otherwise it is weird for this API to change a config

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,16 +30,136 @@
+    _old_size_limit = torch._dynamo.config.cache_size_limit
+
+    try:
+        # We need to turn off the is_fx_tracing_flag. Remove this flag check from dyanmo
+        # once we are confident fx tracing works with dynamo.
+        torch.fx._symbolic_trace._is_fx_tracing_flag = False
```
Sorry, I still don't understand why we are turning this flag to False. You mentioned that making it go to True would make a lot of tests fail. But we shouldn't be regressing behavior, unless they're related to the symbool thing?

ydwu4 replied:
make_fx will 1. turn this flag on and 2. feed FakeTensors and SymBools into torch.compile. I kind of assume this flag can be turned off in cond_compiled and it's not a problem. 

Right now, if we leave this flag alone, we'll get an error for all make_fx tests, which isn't what we want we already supports "real" tracing mode.

zou3519 replied:
Is the problem that in torch.compile, we want `_is_fx_tracing_flag` to be False? If so, I would have expected that we turned that off inside of torch.compile, because this is a make_fx and torch.compile composability issue, rather than doing a one-off solution here

zou3519 commented on test/test_fx.py:
```diff
@@ -329,6 +330,8 @@ def rmatmul_f(x):
         inp = torch.randn(3)
         self.assertEqual(mod(inp), rmatmul_f(inp))
 
+    @skipIfNoDynamoSupport
+    @unittest.expectedFailure
```
Uh, why does this fail?

ydwu4 replied:
This has the SymBool problem.

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1063,7 +1068,7 @@ def true_fn(x):
                     return x.sum() - z.sum() - self.buffer.sum()
 
-                return control_flow.cond(y, true_fn, false_fn, [x])
+                return control_flow.cond(x, true_fn, false_fn, [y])
```
Why did this change? (does it actually matter?)

ydwu4 replied:
Yes, here is a typo where x is tensor[True/False] but y is an interger tensor. Previsouly, we can work for integer tensors. But now, we add an assertion that it must be boolean tensors. This allows us to catch this typo and correct it.

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1037,7 +1037,12 @@ def false_fn(x):
         mod_for_eager = Foo()
-        ref = mod_for_eager(torch.ones(6, 4))
+
+        with self.assertRaisesRegex(
+            torch._dynamo.exc.UserError,
+            r"Can't inplace modify module params/buffers inside HigherOrderOp",
+        ):
```
As I mentioned in the last PR: having our own custom error and message would help surface to the user that "cond needs torch.compile to work, otherwise, it will fail"

ydwu4 replied:
Yeah, I can do it here.

zou3519 commented on test/dynamo/test_export.py:
```diff
@@ -2891,10 +2891,10 @@ def f(x):
         with self.assertRaisesRegex(
-            torch._dynamo.exc.UserError,
-            "Expected 4 arguments",
+            TypeError,
+            r"cond\(\) missing 1 required positional argument: 'operands'",
```
This is a bit suspicious, why did the error message change?

zou3519 replied:
Oh, does it just go through regular Python and that's what raises the TypeError?

ydwu4 replied:
yes, it hits the error at the python wrapper

zou3519 commented on test/dynamo/test_export.py:
```diff
@@ -3009,12 +2999,11 @@ def false_fn(x):
 
         def f_mismatch_return_length(x):
-            return cond(torch.tensor(100), true_fn, false_fn, [x])
+            return cond(torch.tensor(True), true_fn, false_fn, [x])
```
We should... probably just allow values that are coercible to bool. Otherwise it sounds like we might break some things (do we end up breaking anything internal?)

zou3519 replied:
I don't really know if there's a clean way to do this

zou3519 replied:
easiest way is probably just to validate that `pred` is a python number type or a scalar Tensor (we don't need to check the dtype)

ydwu4 replied:
em... I don't know how large chance we'll break internal tests. It also sound good to me to not check the dtype. Maybe let's just change it back to original behavior that we allow boolean casting in predicate

zou3519 commented on test/dynamo/test_export.py:
```diff
@@ -3742,6 +3731,7 @@ def forward(self, x):
 
         self.assertTrue(torch.allclose(m(x), gm(x)))
 
+    @unittest.expectedFailure
```
What's going on here? IIRC this test is important for quantization for 2.1?

ydwu4 replied:
I'll have a check and get back. It has some complicated errors as I remember.

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,16 +30,136 @@
 In order to do this, we need implementations for each of the dispatch keys.
 """
-cond = HigherOrderOperator("cond")
+cond_higher_order_op = HigherOrderOperator("cond")
```
Sounds good!

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1094,6 +1102,7 @@ def false_fn(x):
                     self.assertEqual(num_placeholders, 5)
 
     def test_cond_side_effect_in_one_branches(self):
+        torch._dynamo.reset()
```
You can do this more systematically: in the setup() of the TestCase, just call torch._dynamo.reset()

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,16 +30,130 @@
     reason: str
 
 
+def cond(pred, true_branch, false_branch, operands):
+    r"""
```
true_branch, false_branch -> "true_fn", "false_fn"

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -28,16 +30,130 @@
 
+def cond(pred, true_branch, false_branch, operands):
+    r"""
+    `cond` can logically be seen as implemented as follows::
+
+    def cond(pred, true_branch, false_branch, operands):
+        if pred:
+            return true_branch(*operands)
+        else:
+            return false_branch(*operands)
+
+    Like a normal if statement in Python, users can use it to selectively execute a branch based on
+    the truth value of predicate in eager-mode or torch.compile. What make 'cond' unique is that
+    'cond' is exportable, lowerable: users can use it to capture a data-dependent control flow precisely,
+    preserve it accross the Pytorch stack and compile it in an environemnt without Python.
```
Please change the intro to state the following. Let's clearly state that this feature is a prototype and may not work well outside of export:
```
Conditionally applies ``true_fn`` or ``false_fn``.

``cond`` is structured control flow operator. That is, it is like a Python if-statement, but has limitations on ``true_fn``, ``false_fn``, and ``operands`` that enable it to be capturable using torch.compile and torch.export.

Assuming the constraints on ``cond``'s arguments are met, ``cond`` is equivalent to the following::

    def cond(pred, true_branch, false_branch, operands):
        if pred:
            return true_branch(*operands)
        else:
            return false_branch(*operands)

.. warning::
   cond is a prototype feature in PyTorch, included as a part of the torch.export release. The main limitations are that it may not work in eager-mode PyTorch and you may encounter various failure modes while using it. Please look forward to a more stable implementation in a future version of PyTorch.

Read more about feature classification at: https://pytorch.org/blog/pytorch-feature-classification-changes/#prototype

zou3519 reviewed (approved):
FWIW, cond works pretty reasonably with export and torch.compile. The main issues are that it doesn't really work well with eager-mode PyTorch yet (we have a bunch of PRs that are in flight).

I'm fine with merging this as long as we:
(1) clearly indicate that it is a prototype feature (included with the export release) and document that it might not work well, (2) do some minor cleanup of the things we can do now (e.g. changing the argument names to what they should be)
(3) We should probably allow boolean casting in the predicate. Because otherwise the docstring is kind of wrong :P, and users would expect it (as indicated by the large amount of tests we have that exercised this)
(4) Make sure we don't regress any of the export or compile related tests. That is, test_predispatch_with_for_out_dtype_nested should still pass.

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "Automatically turn on dynamo in cond"


A replacement of https://github.com/pytorch/pytorch/pull/107932.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge
